changelog/unreleased/features/1987-1992--metrics-metatdata.md

```diff
@@ -0,0 +1,7 @@
+Metrics events now optionally contain a metadata field that is a key-value
+mapping of string to string, allowing for finer-grained introspection. For now
+this enables correlation of metrics events and individual queries. A set of new
+metrics for query lookup use this feature to include the query ID. I.e.:
+```
+{"ts": "2021-12-09T09:42:16.797478769","version": "2021.11.18-266-gbb60334346","actor": "meta-index","key": "meta-index.lookup.candidates","value": 42,"query":"DBECE02A-680C-4FB7-A806-69F262B2DB0A"}
```

Comment:
I don't think this example JSON is even correct, because it's missing the metadata field in which the query field should be nested.

Reply:
As explained in the documentation the metadata fields are folded into the top level object for the `file` and `UDS` sinks.

Reply:
Doesn't that add unnecessary complexity?

Reply:
I guess it doesn't hurt to make the outputs identical.

I initially did it this way because the `self-sink` is somewhat unused nowadays and I thought the nesting under `metadata` does little to improve the usability.

Reply:
It's alright to me. No need to change it now.

I think we wanted to get rid of the self-sink anyways, so that may be a good thing to do in the near future. This whole `vast.metrics` thing never really worked out the way we wanted to and it makes writing metrics unnecessarily complex.

Reply:
I aligned it now anyways, so we still have all options in the future.